polyglot-kotlin/src/main/kotlin/org/sonatype/maven/polyglot/kotlin/serialization/ModelScriptWriter.kt

```diff
@@ -0,0 +1,916 @@
+import org.apache.maven.project.MavenProject
+import org.codehaus.plexus.util.xml.Xpp3Dom
+import org.sonatype.maven.polyglot.execute.ExecuteContext
+import org.sonatype.maven.polyglot.kotlin.dsl.addAll
+import org.sonatype.maven.polyglot.kotlin.dsl.addAllNonNull
+import org.sonatype.maven.polyglot.kotlin.dsl.addFirstNonNull
+import org.sonatype.maven.polyglot.kotlin.dsl.cast
+import org.sonatype.maven.polyglot.kotlin.support.anyOf
+import org.sonatype.maven.polyglot.kotlin.support.isNotNull
+import java.io.Writer
+import java.time.LocalDate
+import java.time.LocalTime
+import java.util.*
+
+/**
+ * A `ModelScriptWriter` converts a Maven project model into a Kotlin script capable of producing an identical model.
+ * Instances of this class may be configured with one of two flavors of the Kotlin Model DSL. The most basic flavor,
+ * "block", consists of simply writing callbacks on each model element and using property setters to initialize its
+ * fields. All complex model elements can be configured using this "block" flavor of the DSL. The "block" flavor,
+ * however, does little on its own to reduce the code verbosity of the original XML.
+ *
+ * ```
+ * project {
+ *   groupId = "com.example"
+ *   artifactId = "sample-app
+ *   version = "1.0"
+ *   packaging = "jar"
+ *   name = "Sample Application"
+ *   properties {
+ *     "kotlin.version" to "1.3.11"
+ *   }
+ *   build {
+ *     plugins {
+ *       plugin {
+ *         groupId = "org.jetbrains.kotlin"
+ *         artifactId = "kotlin-maven-plugin"
+ *         version = "\${kotlin.version}"
+ *         executions {
+ *           execution {
+ *             id = "compile"
+ *             phase = "compile"
+ *             goals = listOf("compile")
+ *           }
+ *           execution {
+ *             id = "test-compile"
+ *             phase = "test-compile"
+ *             goals = listOf("test-compile")
+ *           }
+ *         }
+ *         dependencies {
+ *           dependency {
+ *             groupId = "org.jetbrains.kotlin"
+ *             artifactId = "kotlin-maven-allopen"
+ *             version = "\${kotlin.version}"
+ *           }
+ *           dependency {
+ *             groupId = "org.jetbrains.kotlin"
+ *             artifactId = "kotlin-maven-noarg"
+ *             version = "\${kotlin.version}"
+ *           }
+ *         }
+ *         configuration = xml("""
+ *             <configuration>
+ *               <jvmTarget>1.8</jvmTarget>
+ *               <javaParameters>true</javaParameters>
+ *               <compilerPlugins>
+ *                 <plugin>spring</plugin>
+ *                 <plugin>jpa</plugin>
+ *               </compilerPlugins>
+ *             </configuration>""")
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ *
+ * The "mixed" flavor leverages optional parameters on block functions as well as additional extension functions and in
+ * order to minimize the verbosity of specific model elements.
+ *
+ * ```
+ * project("Sample Application") {
```

Comment:
is name required here? because typically in Maven it is not .... maybe the parameter in that constructor should be the typical required ones of GAV

Reply:
No argument is required on the project. When it is supplied, it can be either the "name" or the GAV.